### ISSUE TITLE
Problems with cupy

### ISSUE THREAD

**Issue by tnn111** [nanoporetech/bonito#124]
I've tried all the installation methods I can find and I no matter what, I get the following error. I'm running CUDA 11.2 on an A100 and I can't run cupy-cuda102. Does anyone else have this problem? Better yet, have you found a solution for it?

Thanks, Torben

```
Traceback (most recent call last):
  File "/home/torben/opt/anaconda3/envs/bonito/bin/bonito", line 33, in <module>
    sys.exit(load_entry_point('ont-bonito', 'console_scripts', 'bonito')())
  File "/home/torben/opt/anaconda3/envs/bonito/bin/bonito", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/torben/opt/bonito/bonito/__init__.py", line 3, in <module>
    from bonito.cli import basecaller, evaluate, view, convert, download, pair
  File "/home/torben/opt/bonito/bonito/cli/basecaller.py", line 16, in <module>
    from bonito.fast5 import get_reads, read_chunks
  File "/home/torben/opt/bonito/bonito/fast5.py", line 14, in <module>
    from ont_fast5_api.fast5_interface import get_fast5_file
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/site-packages/ont_fast5_api/__init__.py", line 29, in <module>
    from ont_fast5_api.compression_settings import register_plugin
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/site-packages/ont_fast5_api/compression_settings.py", line 1, in <module>
    import pkg_resources
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3268, in <module>
    def _initialize_master_working_set():
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3251, in _call_aside
    f(*args, **kwargs)
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3280, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/site-packages/pkg_resources/__init__.py", line 582, in _build_master
    ws.require(__requires__)
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/site-packages/pkg_resources/__init__.py", line 899, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/home/torben/opt/anaconda3/envs/bonito/lib/python3.8/site-packages/pkg_resources/__init__.py", line 785, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'cupy-cuda102' distribution was not found and is required by seqdist
```

**Comment by iiSeymour** [comment]
Hey @tnn111 

The requirement for `cupy-cuda102` comes from [seqdist](https://github.com/davidcpage/seqdist) - to use CUDA 11.2 you will need to first install `cupy-cuda112` then checkout and build `seqdist`. However PyTorch is still built against CUDA 10.2 as default but does provide builds against CUDA 11, see https://pytorch.org/get-started/locally/. So unless you want to build PyTorch against CUDA 11.2 yourself I would probably install CUDA 11.0 _(you can install multiple CUDA toolkits side by side)_ and target that. 

See also https://github.com/nanoporetech/bonito/issues/60.

**Comment by russellsmithies** [comment]
Nice to see someone else with the same problems  ;-)
To add to the fun I'm building it in a Singularity container. 
`ont-bonito` runs fine on our other server with V100s but we'd like to get it running on the A100s as well and dropping to 10.x isn't possible as AFAIK A100 support doesn't come in until 11.x 
I might have a go at building PyTorch and cupy against 11.2,

**Comment by russellsmithies** [comment]
Just to follow-up, I have bonito running in a Singularity container with our A100 GPUs and CUDA11.2  :-)
The basic steps are:
- use  `nvidia/cuda:11.2.0-cudnn8-devel-ubuntu20.04` as the base image for the container
- install miniconda
- create a conda env with all requirements for the builds
- build cupy from src (or use `cupy-cuda12`)
- build seqdist from src - but change the `settings.ini` to require `cupy` rather than `cupy-cuda*`
- build pytorch from src - takes about 60 minutes :-(
- build bonito from src - but remove `torch`, `seqdist`, and `numpy` from `requirements.txt`
It took a while to get it all working but basecalling runs at about  2.3E+06 samples per second on a single A100.

**Comment by KasperSkytte** [comment]
@russellsmithies can you perhaps share the singularity definition file?

**Comment by russellsmithies** [comment]
> 
> 
> @russellsmithies can you perhaps share the singularity definition file?

try this @KasperSkytte  https://gist.github.com/russellsmithies/f91a5f62f9219447a147dab7357a3188

**Comment by KasperSkytte** [comment]
Thank you very much @russellsmithies

**Comment by BenjaminSchwessinger** [comment]
Questions along the same issue re cupy-cuda and such. Does bonito and seqdist also work with `cuda10.1` and `cupy-cuda101==9.1.0` or would you recommend to build seqdist from source?

**Comment by russellsmithies** [comment]
> 
> 
> Questions along the same issue re cupy-cuda and such. Does bonito and seqdist also work with `cuda10.1` and `cupy-cuda101==9.1.0` or would you recommend to build seqdist from source?

I had to build because I'm using CUDA 11.2 for our A100 GPUs, but I think if you're using CUDA 10.x you should be able to use the standard builds. The main issue turned out to be pytorch but everything built fairly easily. It's just fairly time-consuming to build.

**Comment by KasperSkytte** [comment]
Maybe helpful to some, but this is a general problem with CUDA that there is a big difference between versions of CUDA and different software tools can require only one or a few specific versions to work properly. I am starting to accept that having multiple versions of CUDA installed on the same machine is necessary, as even containers use CUDA and nvidia drivers on host and not the container. This has solved several issues for me. All you have to do is install whatever versions you need and just update the `LD_LIBRARY_PATH` variable to include either the one you need, or all at once, as tools will likely find the correct version themselves among the different ones available. See https://medium.com/@peterjussi/multicuda-multiple-versions-of-cuda-on-one-machine-4b6ccda6faae

**Comment by russellsmithies** [comment]
We run multiple versions of CUDA because we have V100 and A100 GPUs, and the A100s require CUDA 11.x. and Centos/RHEL 8.x but the V100s are still happy with Centos/RHEL 7.  We've found (mostly) that code compiled or containers built with the newer CUDA are backwards compatible i.e. If I build a Singularity container using CUDA 11.2 then it will run just fine on the V100s, but if I build a container with CUDA 10.2 it won't run on the A100s. I try and package everything into Singularity containers so they're reproducible and portable and it works 99.9% of the time - it's only CUDA that caused issues until we figured out what was going on.
